jdtls.ext/com.microsoft.jdtls.ext.core/src/com/microsoft/jdtls/ext/core/ProjectCommand.java

```diff
@@ -232,23 +234,32 @@ private static boolean exportJarExecution(String mainClass, Classpath[] classpat
     }
 
     public static List<MainClassInfo> getMainClasses(List<Object> arguments, IProgressMonitor monitor) throws Exception {
-        List<PackageNode> projectList = listProjects(arguments, monitor);
-        final List<MainClassInfo> res = new ArrayList<>();
-        List<IJavaElement> searchRoots = new ArrayList<>();
+        List<Object> args = new ArrayList<>(arguments);
+        if (args.size() <= 1) {
+            args.add(Boolean.TRUE);
+        } else {
+            args.set(1, Boolean.TRUE);
+        }
+        List<PackageNode> projectList = listProjects(args, monitor);
         if (projectList.size() == 0) {
-            return res;
+            return Collections.emptyList();
         }
-        for (PackageNode project : projectList) {
+        final List<MainClassInfo> res = new ArrayList<>();
+        List<IJavaProject> javaProjects = new ArrayList<>();
+        for (PackageNode project: projectList) {
             IJavaProject javaProject = PackageCommand.getJavaProject(project.getUri());
-            for (IPackageFragmentRoot packageFragmentRoot : javaProject.getAllPackageFragmentRoots()) {
-                if (!packageFragmentRoot.isArchive()) {
-                    searchRoots.add(packageFragmentRoot);
-                }
+            if (javaProject != null && javaProject.exists()) {
+                javaProjects.add(javaProject);
             }
         }
-        IJavaSearchScope scope = SearchEngine.createJavaSearchScope(searchRoots.toArray(new IJavaElement[0]));
-        SearchPattern pattern = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
-                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_EXACT_MATCH | SearchPattern.R_CASE_SENSITIVE);
+        int includeMask = IJavaSearchScope.SOURCES;
+        IJavaSearchScope scope = SearchEngine.createJavaSearchScope(javaProjects.toArray(new IJavaProject[0]),
+                    includeMask);
+        SearchPattern pattern1 = SearchPattern.createPattern("main(String[]) void", IJavaSearchConstants.METHOD,
+                IJavaSearchConstants.DECLARATIONS, SearchPattern.R_CASE_SENSITIVE | SearchPattern.R_EXACT_MATCH);
+        SearchPattern pattern2 = SearchPattern.createPattern("main() void", IJavaSearchConstants.METHOD,
```

Comment:
why add "main() void", because this does not conform to the standard definition of the main method in Java.